ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockTransactionSelectorTest.java

```diff
@@ -405,14 +438,119 @@ public void transactionSelectionStopsWhenSufficientBlockOccupancyIsReached() {
+    final BlockTransactionSelector selector =
+        createBlockSelector(
+            transactionProcessor,
+            blockHeader,
+            Wei.ZERO,
+            miningBeneficiary,
+            Wei.ZERO,
+            MIN_OCCUPANCY_100_PERCENT);
+
+    final long minTxGasCost = getGasCalculator().getMinimumTransactionCost();
+
+    // Add 4 transactions to the Pending Transactions
+    // 1) 90% of block (selected)
+    // 2) 90% of block (skipped since too large)
+    // 3) enough gas to only leave space for a transaction with the min gas cost (selected)
+    // 4) min gas cost (selected and 100% block gas used)
+    // 5) min gas cost (not selected since selection stopped after tx 4)
+    // NOTE - PendingTransactions outputs these in nonce order
+
+    final long gasLimit0 = (long) (blockHeader.getGasLimit() * 0.9);
+    final long gasLimit1 = (long) (blockHeader.getGasLimit() * 0.9);
+    final long gasLimit2 = blockHeader.getGasLimit() - gasLimit0 - minTxGasCost;
+    final long gasLimit3 = minTxGasCost;
+    final long gasLimit4 = minTxGasCost;
+
+    final TransactionTestFixture txTestFixture = new TransactionTestFixture();
+    final List<Transaction> transactionsToInject = Lists.newArrayList();
+    transactionsToInject.add(txTestFixture.gasLimit(gasLimit0).nonce(0).createTransaction(keyPair));
+    transactionsToInject.add(txTestFixture.gasLimit(gasLimit1).nonce(1).createTransaction(keyPair));
+    transactionsToInject.add(txTestFixture.gasLimit(gasLimit2).nonce(2).createTransaction(keyPair));
+    transactionsToInject.add(txTestFixture.gasLimit(gasLimit3).nonce(3).createTransaction(keyPair));
+    transactionsToInject.add(txTestFixture.gasLimit(gasLimit4).nonce(4).createTransaction(keyPair));
+
+    for (final Transaction tx : transactionsToInject) {
+      pendingTransactions.addRemoteTransaction(tx, Optional.empty());
+      ensureTransactionIsValid(tx);
+    }
+
+    final BlockTransactionSelector.TransactionSelectionResults results =
+        selector.buildTransactionListForBlock();
+
+    assertThat(results.getTransactions())
+        .containsExactly(
+            transactionsToInject.get(0), transactionsToInject.get(2), transactionsToInject.get(3));
+    assertThat(results.getCumulativeGasUsed()).isEqualTo(blockHeader.getGasLimit());
+  }
+
+  @Test
+  public void transactionSelectionStopsWhenRemainingGasIsNotEnoughForAnyMoreTransaction() {
+    final ProcessableBlockHeader blockHeader = createBlock(3_000_000);
+
+    final Address miningBeneficiary = AddressHelpers.ofValue(1);
+    final BlockTransactionSelector selector =
+        createBlockSelector(
+            transactionProcessor,
+            blockHeader,
+            Wei.ZERO,
+            miningBeneficiary,
+            Wei.ZERO,
+            MIN_OCCUPANCY_100_PERCENT);
+
+    final long minTxGasCost = getGasCalculator().getMinimumTransactionCost();
+
+    // Add 4 transactions to the Pending Transactions
+    // 1) 90% of block (selected)
```

Comment:
nit - if you numbered these 0..4 it would match the code

Reply:
good catch, will renumber